data/trie/trieStorageManager.go

```diff
@@ -0,0 +1,326 @@
+		dbEvictionWaitingList: tsm.dbEvictionWaitingList,
+	}
+}
+
+// Prune removes the given hash from db
+func (tsm *trieStorageManager) Prune(rootHash []byte) error {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	if tsm.snapshotsBuffer.len() != 0 {
+		tsm.pruningBuffer = append(tsm.pruningBuffer, rootHash)
+		return nil
+	}
+
+	err := tsm.removeFromDb(rootHash)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CancelPrune removes the given hash from the eviction waiting list
+func (tsm *trieStorageManager) CancelPrune(rootHash []byte) {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	_, _ = tsm.dbEvictionWaitingList.Evict(rootHash)
+}
+
+func (tsm *trieStorageManager) removeFromDb(hash []byte) error {
+	hashes, err := tsm.dbEvictionWaitingList.Evict(hash)
+	if err != nil {
+		return err
+	}
+
+	for i := range hashes {
+		err = tsm.db.Remove(hashes[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// MarkForEviction adds the given hashes in the eviction waiting list at the provided key
+func (tsm *trieStorageManager) MarkForEviction(root []byte, hashes [][]byte) error {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	return tsm.dbEvictionWaitingList.Put(root, hashes)
+}
+
+// GetDbThatContainsHash returns the database that contains the given hash
+func (tsm *trieStorageManager) GetDbThatContainsHash(rootHash []byte) data.DBWriteCacher {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	_, err := tsm.db.Get(rootHash)
+
+	hashPresent := err == nil
+	if hashPresent {
+		return tsm.db
+	}
+
+	for i := range tsm.snapshots {
+		_, err = tsm.snapshots[i].Get(rootHash)
+
+		hashPresent = err == nil
+		if hashPresent {
+			return tsm.snapshots[i]
+		}
+	}
+
+	return nil
+}
+
+// Snapshot creates a new snapshot, or if there is another snapshot or checkpoint in progress,
+// it adds this snapshot in the queue.
+func (tsm *trieStorageManager) Snapshot(rootHash []byte, msh marshal.Marshalizer, hsh hashing.Hasher) {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	tsm.snapshotsBuffer.add(rootHash, true)
+	if tsm.snapshotsBuffer.len() > 1 {
+		return
+	}
+
+	go tsm.snapshot(msh, hsh)
+}
+
+// Checkpoint creates a new checkpoint, or if there is another snapshot or checkpoint in progress,
+// it adds this checkpoint in the queue. The checkpoint operation creates a new snapshot file
+// only if there was no snapshot done prior to this
+func (tsm *trieStorageManager) Checkpoint(rootHash []byte, msh marshal.Marshalizer, hsh hashing.Hasher) {
+	tsm.storageOperationMutex.Lock()
+	defer tsm.storageOperationMutex.Unlock()
+
+	tsm.snapshotsBuffer.add(rootHash, false)
+	if tsm.snapshotsBuffer.len() > 1 {
+		return
+	}
+
+	go tsm.snapshot(msh, hsh)
+}
+
+func (tsm *trieStorageManager) snapshot(msh marshal.Marshalizer, hsh hashing.Hasher) {
+	var keys [][]byte
+	for {
```

Comment:
it would be easier to understand if you add a condition to this `for`. for example:
``` GOLANG
isSnapshotsBufferEmpty := false
for !isSnapshotsBufferEmpty{
....
  isSnapshotsBufferEmpty = tsm.snapshotsBuffer.len() == 0
....
}
```

Reply:
Done.